dm/syncer/data_validator.go

```diff
@@ -138,52 +224,167 @@ func (v *DataValidator) fillResult(err error, needLock bool) {
 		defer v.Unlock()
 	}
 
-	var errs []*pb.ProcessError
-	if utils.IsContextCanceledError(err) {
-		v.L.Info("filter out context cancelled error", log.ShortError(err))
-	} else {
-		errs = append(errs, unit.NewProcessError(err))
-	}
-
+	// todo: if error, validation is stopped( and cancelled), and we may receive a cancelled error.
+	// todo: do we need this cancel field?
 	isCanceled := false
 	select {
 	case <-v.ctx.Done():
 		isCanceled = true
 	default:
 	}
+	v.result.IsCanceled = isCanceled
+
+	if utils.IsContextCanceledError(err) {
+		v.L.Info("filter out context cancelled error", log.ShortError(err))
+	} else {
+		v.result.Errors = append(v.result.Errors, unit.NewProcessError(err))
+	}
+}
+
+func (v *DataValidator) errorProcessRoutine() {
+	defer v.errProcessWg.Done()
+	for err := range v.errChan {
+		v.fillResult(err, true)
+
+		if errors.Cause(err) != context.Canceled {
+			// todo: need a better way to handle err(auto resuming on some error, etc.)
+			v.stopInner()
+		}
+	}
+}
+
+func (v *DataValidator) waitSyncerSynced(currLoc binlog.Location) error {
+	syncLoc := v.syncer.checkpoint.FlushedGlobalPoint()
+	cmp := binlog.CompareLocation(currLoc, syncLoc, v.cfg.EnableGTID)
+	if cmp <= 0 {
+		return nil
+	}
+
+	fired := false
+	v.waitSyncerTimer.Reset(checkInterval)
+	defer func() {
+		if !fired {
+			if !v.waitSyncerTimer.Stop() {
+				<-v.waitSyncerTimer.C
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-v.ctx.Done():
+			return v.ctx.Err()
+		case <-v.waitSyncerTimer.C:
+			fired = true
+			syncLoc = v.syncer.checkpoint.FlushedGlobalPoint()
+			cmp = binlog.CompareLocation(currLoc, syncLoc, v.cfg.EnableGTID)
+			if cmp <= 0 {
+				return nil
+			}
+			v.L.Info("wait syncer synced", zap.Reflect("loc", currLoc))
+			v.waitSyncerTimer.Reset(checkInterval)
+			fired = false
+		}
+	}
+}
 
-	v.result = pb.ProcessResult{
-		IsCanceled: isCanceled,
-		Errors:     errs,
+func (v *DataValidator) waitSyncerRunning() error {
+	if v.syncer.IsRunning() {
+		return nil
+	}
+	for {
+		select {
+		case <-v.ctx.Done():
+			return v.ctx.Err()
+		case <-time.After(checkInterval):
+			if v.syncer.IsRunning() {
+				return nil
+			}
+		}
 	}
 }
 
+// doValidate: runs in a separate goroutine.
 func (v *DataValidator) doValidate() {
-	tctx := tcontext.NewContext(v.ctx, v.L)
-	err := v.streamerController.Start(tctx, lastLocationForTest)
-	if err != nil {
-		v.fillResult(terror.Annotate(err, "fail to restart streamer controller"), true)
+	if err := v.waitSyncerRunning(); err != nil {
+		v.errChan <- terror.Annotate(err, "failed to wait syncer running")
 		return
 	}
 
+	// todo: syncer may change replication location(start from timestamp, sharding resync), how validator react?
```

Comment:
as we have discussed, syncer may start from a timestamp, but since the task need to be stopped, so this case will not happen.

left a todo, to handle the case the case validator may get a later location when validator starts on task start.